testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/UserStorageOTPTest.java

```diff
@@ -0,0 +1,230 @@
+
+    private String dummyProviderId;
+    private String testUserId;
+
+
+    @Page
+    protected LoginPage loginPage;
+
+    @Page
+    protected LoginTotpPage loginTotpPage;
+
+    @Page
+    protected LoginConfigTotpPage loginConfigTotpPage;
+
+    @Page
+    protected AppPage appPage;
+
+    protected TimeBasedOTP totp = new TimeBasedOTP();
+
+
+
+    @Override
+    public void configureTestRealm(RealmRepresentation testRealm) {
+
+    }
+
+    @Before
+    public void addProvidersBeforeTest() throws URISyntaxException, IOException {
+        ComponentRepresentation dummyProvider = new ComponentRepresentation();
+        dummyProvider.setName("dummy");
+        dummyProvider.setId(DummyUserFederationProviderFactory.PROVIDER_NAME);
+        dummyProvider.setProviderId(DummyUserFederationProviderFactory.PROVIDER_NAME);
+        dummyProvider.setProviderType(UserStorageProvider.class.getName());
+        dummyProvider.setConfig(new MultivaluedHashMap<>());
+        dummyProvider.getConfig().putSingle("priority", Integer.toString(0));
+        dummyProvider.getConfig().putSingle(IMPORT_ENABLED, Boolean.toString(false));
+
+        dummyProviderId = addComponent(testRealm(), getCleanup(), dummyProvider);
+
+        UserRepresentation user = UserBuilder.create()
+                .username("test-user")
+                .email("test-user@something.org")
+                .build();
+        String testUserId = ApiUtil.createUserWithAdminClient(testRealm(), user);
+
+        getCleanup().addUserId(testUserId);
+    }
+
+
+    @Test
+    public void testCredentialsThroughRESTAPI() {
+        // Test that test-user has federation link on him
+        UserResource user = ApiUtil.findUserByUsernameId(testRealm(), "test-user");
+        Assert.assertEquals(DummyUserFederationProviderFactory.PROVIDER_NAME, user.toRepresentation().getFederationLink());
+
+        // Test that both "password" and "otp" are configured for the test-user
+        List<String> userStorageCredentialTypes = user.getConfiguredUserStorageCredentialTypes();
+        Assert.assertNames(userStorageCredentialTypes, PasswordCredentialModel.TYPE, OTPCredentialModel.TYPE);
+    }
+
+
+    @Test
+    public void testAuthentication() {
+        // Test that user is required to provide OTP credential during authentication
+        loginPage.open();
+        loginPage.login("test-user", DummyUserFederationProvider.HARDCODED_PASSWORD);
+
+        loginTotpPage.assertCurrent();
+
+        loginTotpPage.login("654321");
+        loginTotpPage.assertCurrent();
+        Assert.assertEquals("Invalid authenticator code.", loginPage.getError());
+
+        loginTotpPage.login(DummyUserFederationProvider.HARDCODED_OTP);
+
+        appPage.assertCurrent();
+        Assert.assertEquals(AppPage.RequestType.AUTH_RESPONSE, appPage.getRequestType());
+        Assert.assertNotNull(oauth.getCurrentQuery().get(OAuth2Constants.CODE));
+    }
+
+
+    private void authenticateUser() {
```

Comment:
This looks redundant.

Reply:
Thanks, I've removed method "authenticateUser" and also did few very minor cleanups in the UserStorageOTPTest. Used separate commit for that, so it's easier to re-review